mlflow/langchain/__init__.py

```diff
@@ -151,6 +163,48 @@ def save_model(
 
                      .. Note:: Experimental: This parameter may change or be removed in a future
                                              release without warning.
+    :param loader_fn: A function that's required for models containing objects that aren't natively
+                      serialized by LangChain.
+                      This function takes a string `persist_dir` as an argument and returns the
+                      specific object that the model needs. Depending on the model,
+                      this could be a retriever, vectorstore, requests_wrapper, embeddings, or
+                      database. For RetrievalQA models, the object is a
+                      (`retriever <https://python.langchain.com/docs/modules/data_connection/retrievers/>`_).
+                      For APIChain models, it's a
+                      (`requests_wrapper <https://python.langchain.com/docs/modules/agents/tools/integrations/requests>`_).
+                      For HypotheticalDocumentEmbedder models, it's an
+                      (`embeddings <https://python.langchain.com/docs/modules/data_connection/text_embedding/>`_).
+                      For SQLDatabaseChain models, it's a
+                      (`database <https://python.langchain.com/docs/modules/agents/toolkits/sql_database>`_).
+    :param persist_dir: The directory where the object is stored. The `loader_fn`
+                        takes this string as the argument to load the object.
+                        This is optional for models containing objects that aren't natively
+                        serialized by LangChain. MLflow logs the content in this directory as
+                        artifacts in the subdirectory named `persist_dir_data`.
+
+                        Here is the code snippet for logging a RetrievalQA chain with `loader_fn`
+                        and `persist_dir`:
+
+                        .. code-block:: python
+
+                            qa = RetrievalQA.from_llm(llm=OpenAI(), retriever=db.as_retriever())
+
+
+                            def load_retriever(persist_directory):
+                                embeddings = OpenAIEmbeddings()
+                                vectorstore = FAISS.load_local(persist_directory, embeddings)
+                                return vectorstore.as_retriever()
+
+
+                            with mlflow.start_run() as run:
+                                logged_model = mlflow.langchain.log_model(
+                                    qa,
+                                    artifact_path="retrieval_qa",
+                                    loader_fn=load_retriever,
+                                    persist_dir=persist_dir,
+                                )
```

Comment:
Do we need to log the module that defines the loader function as well to make it possible to load the model on a different machine/environment that doesn't have the module?

Reply:
I don't know whether there is an easy way to do this. I think cloudpickle should support most of the cases.
For the cases where cloudpickle does not support, users can specify the modules needed in the pip requirements manually.

If we implement the module logging, then we need to consider logging multiple modules if the current module uses relative imports: `import ..another_module` and things can get pretty complex.

What do you think?

Reply:
let's you have a script named `a.py` and the loader function is that file. You log a model on your laptop and serve it on a different machine. Does this work?

Reply:
Saw this:

https://github.com/cloudpipe/cloudpickle#overriding-pickles-serialization-mechanism-for-importable-constructs

> An important difference between cloudpickle and pickle is that cloudpickle can serialize a function or class by value

Reply:
> let's you have a script named a.py and the loader function is that file. You log a model on your laptop and serve it on a different machine. Does this work?

It depends on where do you call the log_model (essentially cloudpickle.dump()):
1. If you call cloudpickle.dump() right after the function definition (like I did in the test), the function is serialized by value.
2. If you call cloudpickle.dump() elsewhere, for example, in another function, or in another python repl, then the loader_function will be serialized by reference.
I got the above understanding empirically.